Make all networking code mockable

vasild opened the pull request:
_This is a roadmap PR. It can be merged, but it can also be split into separate PRs and to get proper thorough review it is split._

Add wrapper methods to the syscalls `accept()`, `setsockopt()`, `getsockname()`, `bind()`, `listen()`  in the [`Sock`](https://github.com/bitcoin/bitcoin/blob/eb9a1fe03779bf05062b70f14190cb23ff42b46f/src/util/sock.h#L25) class (e.g. `Sock::Accept()`). Those methods can be overriden (mocked) by unit tests ([existent example in `master`](https://github.com/bitcoin/bitcoin/blob/eb9a1fe03779bf05062b70f14190cb23ff42b46f/src/test/util/net.h#L75)) and by fuzz tests ([existent example in `master`](https://github.com/bitcoin/bitcoin/blob/eb9a1fe03779bf05062b70f14190cb23ff42b46f/src/test/fuzz/util.h#L561)).

Change everybody to use [`Sock`](https://github.com/bitcoin/bitcoin/blob/eb9a1fe03779bf05062b70f14190cb23ff42b46f/src/util/sock.h#L25) instead of [`SOCKET`](https://github.com/bitcoin/bitcoin/blob/eb9a1fe03779bf05062b70f14190cb23ff42b46f/src/compat.h#L41).

Move the functionality of [`CConnman::SocketEvents()`](https://github.com/bitcoin/bitcoin/blob/eb9a1fe03779bf05062b70f14190cb23ff42b46f/src/net.cpp#L1334) to a mockable method of the [`Sock`](https://github.com/bitcoin/bitcoin/blob/eb9a1fe03779bf05062b70f14190cb23ff42b46f/src/util/sock.h#L25) class.

### Already merged pieces of this PR
* [x] https://github.com/bitcoin/bitcoin/pull/21943
* [x] https://github.com/bitcoin/bitcoin/pull/23601
* [x] https://github.com/bitcoin/bitcoin/pull/21879
* [x] https://github.com/bitcoin/bitcoin/pull/23604
* [x] https://github.com/bitcoin/bitcoin/pull/24357
* [x] https://github.com/bitcoin/bitcoin/pull/24356
* [x] https://github.com/bitcoin/bitcoin/pull/25428
* [x] https://github.com/bitcoin/bitcoin/pull/25426
* [x] https://github.com/bitcoin/bitcoin/pull/24378
* [x] https://github.com/bitcoin/bitcoin/pull/25421
* [x] https://github.com/bitcoin/bitcoin/pull/26312

### Current pieces of this for review

* [ ] https://github.com/bitcoin/bitcoin/pull/28584

vasild committed:
fuzz: add CConnman::OpenNetworkConnection() to the tests

Now that all network calls done by `CConnman::OpenNetworkConnection()`
are done via `Sock` they can be redirected (mocked) to `FuzzedSocket`
for testing.

vasild committed:
fuzz: add CConnman::CreateNodeFromAcceptedSocket() to the tests

vasild committed:
fuzz: add CConnman::InitBinds() to the tests

vasild committed:
net: remove now unnecessary Sock::Get()

`Sock::Get()` was used only in `sock.{cpp,h}`. Remove it and access
`Sock::m_socket` directly.

Unit tests that used `Get()` to test for equality still verify that the
behavior is correct, indirectly, by testing whether the socket is closed
or not.

vasild committed:
fuzz: add CConnman::SocketHandler() to the tests

practicalswift commented:
Strong Concept ACK: mockable means fuzzable.

jonatack commented:
Concept ACK

promag commented on src/net.h:
```diff
@@ -1267,6 +1267,36 @@ class CConnman
+        std::vector<CNode*> m_nodes_copy;
+
+    private:
+        CConnman* m_connman;
```
96bcc72b5dc4f8ba5654d8ddeab844688507f1ce

nit, * const

promag replied:
I did though of that but then the caller inits with `{*this}`?

vasild replied:
After reading
https://github.com/isocpp/CppCoreGuidelines/issues/222 and
https://github.com/isocpp/CppCoreGuidelines/issues/1707
I decided to use a reference instead of a pointer because the member cannot/must not be nullable.

promag commented on src/net.h:
```diff
@@ -1267,6 +1267,36 @@ class CConnman
+            }
+        }
+
+        std::vector<CNode*> m_nodes_copy;
```
96bcc72b5dc4f8ba5654d8ddeab844688507f1ce

make private and add const getter `const std::vector<CNode*>& nodes() const { return m_nodes_copy; }`

vasild replied:
Done.

promag reviewed (commented):
Concept ACK.

promag commented:
@vasild I think 96bcc72 and refactors around it could easily be merged on its own.

vasild commented:
`0256cfd34...e68bc633f`: address review suggestions

vasild commented:
> @vasild I think [`96bcc72`](https://github.com/bitcoin/bitcoin/commit/96bcc72b5dc4f8ba5654d8ddeab844688507f1ce) and refactors around it could easily be merged on its own.

Right! Submitted in https://github.com/bitcoin/bitcoin/pull/21943.

jnewbery commented:
Very nice. Strong concept ACK.

vasild commented:
`e68bc633f...5772edb36`: add an explicit commit that shows we intentionally don't process newly accepted nodes, as discussed in https://github.com/bitcoin/bitcoin/pull/21943#discussion_r632150326

vasild commented:
`5772edb...ec5f887`: replicate changes to https://github.com/bitcoin/bitcoin/pull/21943

vasild commented:
`ec5f887543...a0b00304a8`: rebase due to conflicts

vasild commented:
`a0b00304a8...6bed261aaf`: rebase due to conflicts, follow the trend from https://github.com/bitcoin/bitcoin/pull/21659 and flag newly added `Sock` methods with `[[nodiscard]]`

vasild committed:
i2p: avoid using Sock::Get() for checking for a valid socket

Peeking at the underlying socket file descriptor of `Sock` and checkig
if it is `INVALID_SOCKET` is bad encapsulation and stands in the way of
testing/mocking/fuzzing.

Instead use an empty unique_ptr to denote that there is no valid socket.

vasild committed:
net: don't check if the socket is valid in GetBindAddress()

The socket is always valid (the underlying file descriptor is not
`INVALID_SOCKET`) when `GetBindAddress()` is called.

vasild committed:
net: don't check if the socket is valid in ConnectSocketDirectly()

The socket is always valid (the underlying file descriptor is not
`INVALID_SOCKET`) when `ConnectSocketDirectly()` is called.

vasild committed:
net: remove Sock default constructor, it's not necessary

vasild commented:
`6bed261aaf...272ecad4bf`:
* Rebase
* Fix https://github.com/bitcoin/bitcoin/issues/21744
* To denote an invalid socket: use an empty `shared_ptr`/`unique_ptr` to `Sock` (that owns `nullptr`) instead of a `Sock` object that contains `m_socket` equal to `INVALID_SOCKET`. This way high-level callers never need to lookup the internal file descriptor of `Sock`.

vasild commented:
`272ecad4bf...2833c71862`: fix a compiler warning

vasild commented:
`2833c71862...f029732b34`: rebase due to conflicts

vasild commented:
`f029732b34...379e9e7279`: rebase due to conflicts

vasild commented:
`379e9e7279...cdf537545a`: during fuzzing, mock also DNS lookup; this is necessary since we now fuzz high level code which may end up doing DNS lookups, for example: `CConnman::OpenNetworkConnection() -> CConnman::ConnectNode() -> Lookup()`

jamesob commented:
Concept ACK, excited to dig into this.

vasild commented:
`cdf537545a...c98c54e3e3`: rebase due to conflicts

vasild commented:
`c98c54e3e3...f8d0ae7e5d`: rebase due to conflicts and address review suggestions from child PRs: https://github.com/bitcoin/bitcoin/pull/21879#issuecomment-928928790, https://github.com/bitcoin/bitcoin/pull/21943#issuecomment-928930542.

vasild commented:
`f8d0ae7e5d...39476dc354`: rebase and pick minor changes from child PR https://github.com/bitcoin/bitcoin/pull/21943#issuecomment-929201658.

vasild commented:
`94f8543cda...3673db7c0a`: rebase due to conflicts

vasild commented:
`3673db7c0a...ff3d6a5937`: pick changes from child PR: https://github.com/bitcoin/bitcoin/pull/21943#issuecomment-966969216

vasild commented:
`ff3d6a5937...63280ff095`: rebase due to conflicts and pick changes from child PR: https://github.com/bitcoin/bitcoin/pull/21943#issuecomment-972830829

vasild commented:
`63280ff095...8fa0821747`: rebase due to conflicts. #21943 which was part of this PR was merged, thus remove its commits from this PR.

vasild commented:
`8fa0821747...9dd36c4a72`: rebase due to conflicts

vasild commented:
`9dd36c4a72...ecd2afdbb8`: rebase due to conflicts

vasild commented:
`ecd2afdbb8...7e74e1f959`: rebase due to conflicts

josibake commented:
Concept ACK

I see two open PR's listed in the description as ready for review; are those the best places to start with reviewing this?

vasild commented:
@josibake, yes. https://github.com/bitcoin/bitcoin/pull/21879 and https://github.com/bitcoin/bitcoin/pull/23604 contain the first commits from this PR. And if you are curious what's coming afterwards - browse the subsequent commits in this PR. Thanks!

vasild commented:
`7e74e1f959...cae35129e6`: rebase due to conflicts; remove 3 commits from this PR because they got merged via https://github.com/bitcoin/bitcoin/pull/21879

PastaPastaPasta commented on src/net.cpp:
```diff
@@ -385,17 +385,15 @@ bool CConnman::CheckIncomingNonce(uint64_t nonce)
-        } else {
-            LogPrint(BCLog::NET, "Warning: getsockname failed\n");
-        }
+    if (!sock.GetSockName((struct sockaddr*)&sockaddr_bind, &sockaddr_bind_len)) {
+        addr_bind.SetSockAddr((const struct sockaddr*)&sockaddr_bind);
```
please replace these c-style casts with reinterpret_casts

vasild replied:
The cast was not added in this PR. I am leaving it as is because it is unrelated to the change and without it, it is easier to see that only `getsockname` has been changed to `sock.GetSockName`.

PastaPastaPasta commented on src/net.cpp:
```diff
@@ -2358,22 +2236,36 @@ bool CConnman::BindListenPort(const CService& addrBind, bilingual_str& strError,
     // Allow binding if the port is still in TIME_WAIT state after
     // the program was closed and restarted.
-    setsockopt(sock->Get(), SOL_SOCKET, SO_REUSEADDR, (sockopt_arg_type)&nOne, sizeof(int));
+    if (sock->SetSockOpt(SOL_SOCKET, SO_REUSEADDR, (sockopt_arg_type)&nOne, sizeof(int)) == SOCKET_ERROR) {
```
please replace this c-style cast with a reinterpret_cast (I think)

vasild replied:
See https://github.com/bitcoin/bitcoin/pull/21878#discussion_r804795139

PastaPastaPasta commented on src/net.cpp:
```diff
@@ -2358,22 +2236,36 @@ bool CConnman::BindListenPort(const CService& addrBind, bilingual_str& strError,
     if (addrBind.IsIPv6()) {
 #ifdef IPV6_V6ONLY
-        setsockopt(sock->Get(), IPPROTO_IPV6, IPV6_V6ONLY, (sockopt_arg_type)&nOne, sizeof(int));
+        if (sock->SetSockOpt(IPPROTO_IPV6, IPV6_V6ONLY, (sockopt_arg_type)&nOne, sizeof(int)) == SOCKET_ERROR) {
```
same (reinterpret_cast)

vasild replied:
See https://github.com/bitcoin/bitcoin/pull/21878#discussion_r804795139

PastaPastaPasta commented on src/net.cpp:
```diff
@@ -2358,22 +2236,36 @@ bool CConnman::BindListenPort(const CService& addrBind, bilingual_str& strError,
 #ifdef WIN32
         int nProtLevel = PROTECTION_LEVEL_UNRESTRICTED;
-        setsockopt(sock->Get(), IPPROTO_IPV6, IPV6_PROTECTION_LEVEL, (const char*)&nProtLevel, sizeof(int));
+        if (sock->SetSockOpt(IPPROTO_IPV6, IPV6_PROTECTION_LEVEL, (const char*)&nProtLevel, sizeof(int)) == SOCKET_ERROR) {
```
same

vasild replied:
See https://github.com/bitcoin/bitcoin/pull/21878#discussion_r804795139

PastaPastaPasta commented on src/netbase.cpp:
```diff
@@ -511,19 +511,23 @@ std::unique_ptr<Sock> CreateSockTCP(const CService& address_family)
     // Set the no-sigpipe option on the socket for BSD systems, other UNIXes
     // should use the MSG_NOSIGNAL flag for every send.
-    setsockopt(hSocket, SOL_SOCKET, SO_NOSIGPIPE, (void*)&set, sizeof(int));
+    if (sock->SetSockOpt(SOL_SOCKET, SO_NOSIGPIPE, (void*)&set, sizeof(int)) == SOCKET_ERROR) {
```
reinterpret_cast

vasild replied:
See https://github.com/bitcoin/bitcoin/pull/21878#discussion_r804795139

PastaPastaPasta commented on src/netbase.cpp:
```diff
@@ -543,10 +547,6 @@ bool ConnectSocketDirectly(const CService &addrConnect, const Sock& sock, int nT
-        LogPrintf("Cannot connect to %s: invalid socket\n", addrConnect.ToString());
-        return false;
-    }
     if (!addrConnect.GetSockAddr((struct sockaddr*)&sockaddr, &len)) {
```
reinterpret_cast

vasild replied:
See https://github.com/bitcoin/bitcoin/pull/21878#discussion_r804795139

vasild commented:
`cae35129e6...4a26bcd3fe`: rebase due to conflicts, remove 3 commits from this PR because they got merged via https://github.com/bitcoin/bitcoin/pull/23604

vasild commented:
`4a26bcd3fe...bca585381d`: rebase and minor improvements after self-review and before chopping off the next piece into its own PR: rename `Sock::WaitEvents` to `Sock::Events` and make it possible to define an instance of it without specifying the "occurred" event, which does not make sense anyway.

vasild commented:
`bca585381d...b73dcbc24e`: rebase due to conflicts and pick changes from child PRs

vasild commented:
`b73dcbc24e...a8f889d5e2`: rebase due to conflicts

vasild commented:
`a8f889d5e2...e834d610ff`: rebase due to conflicts

vasild commented:
`e834d610ff...f3e35c4d6c`: rebase due to conflicts and pick changes from child PR https://github.com/bitcoin/bitcoin/pull/24356#issuecomment-1131682426

vasild commented:
`f3e35c4d6c...0aae576f68`: rebase due to conflicts

vasild commented:
`0aae576f68...6077fac5ef`: rebase due to conflicts

vasild commented:
`6077fac5ef...82a6d4576e`: rebase due to conflicts

vasild commented:
`82a6d4576e...a6424d5023`: rebase due to conflicts

vasild commented:
`a6424d5023...12f848b5cd`: rebase due to conflicts

vasild commented:
`12f848b5cd...f07b89e917`: rebase due to conflicts

achow101 commented:
Should this be marked as draft since it is a parent PR?

jonatack commented:
@vasild this may be able to be closed now that all of the child PRs have been merged.

fanquake commented:
> vasild this may be able to be closed now that all of the child PRs have been merged.

Given that the majority of commits in this pull haven't been merged, I don't think so.

vasild commented:
`f07b89e917...9a27807453`: rebase due to conflicts

> Should this be marked as draft since it is a parent PR?

Could be, I am not sure. The reason I did not make it a draft is because it can be merged as it is - there is no danger that something bad will happen if it gets merged (assuming there are reviewers who ACK it). The reason I split it into smaller PRs is solely to ease review. Should I make it draft?

> @vasild this may be able to be closed now that all of the child PRs have been merged.

> Given that the majority of commits in this pull haven't been merged, I don't think so.

Right, there are two pieces left - https://github.com/bitcoin/bitcoin/pull/26312 which I just opened and the tests themselves (which depend on that PR, thus no separate PR for them yet).

vasild commented:
`9a27807453...77ab5e8e98`: rebase due to conflicts

vasild commented:
`77ab5e8e98...b497200c7a`: rebase due to conflicts

vasild commented:
`b497200c7a...80d9dfd0f0`: rebase due to conflicts

vasild commented:
All the functional bits of this are now merged. Thank you all!

This enables extending the test framework to be able to call _any_ method, without worries that it may end up doing OS syscalls (try to create sockets, open connections, etc). A PR that does some of that is in https://github.com/bitcoin/bitcoin/pull/28584 (= last 4 commits from this PR).

Closing as complete.
